File: DSA/Pranvkumar_590011587/Day25_Question1.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

#  Problem 1: Palindrome Linked List
def isPalindrome(head: ListNode) -> bool:
    if not head or not head.next:
        return True
    # Step 1: Find middle
    slow = fast = head
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
    # Step 2: Reverse second half
    prev = None
    while slow:
        nxt = slow.next
        slow.next = prev
        prev = slow
        slow = nxt
    # Step 3: Compare both halves
    left, right = head, prev
    while right:
        if left.val != right.val:
            return False
        left = left.next
        right = right.next
    return True

def build_linked_list(values):
    dummy = current = ListNode(0)
    for v in values:
        current.next = ListNode(v)
        current = current.next
    return dummy.next

def print_linked_list(head):
    vals = []
    while head:
        vals.append(str(head.val))
        head = head.next
    print(" -> ".join(vals))

if __name__ == "__main__":
    print("---- Problem 1: Palindrome Check ----")
    list = build_linked_list([1, 2, 1, 1, 2, 1])
    print("Input: 1 -> 2 -> 1 -> 1 -> 2 -> 1")
    print("Output:", isPalindrome(list))  

    list1 = build_linked_list([1, 2, 3, 4])
    print("\nInput: 1 -> 2 -> 3 -> 4")
    print("Output:", isPalindrome(list1)) 

File: DSA/PriyankaSingh_590016827/day25question1.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

// Function to create a new node
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to insert node at end
void append(struct Node** headRef, int data) {
    struct Node* newNode = createNode(data);
    if (*headRef == NULL) {
        *headRef = newNode;
        return;
    }
    struct Node* temp = *headRef;
    while (temp->next != NULL)
        temp = temp->next;
    temp->next = newNode;
}

// Function to reverse a linked list
struct Node* reverse(struct Node* head) {
    struct Node* prev = NULL;
    struct Node* current = head;
    struct Node* next = NULL;
    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    return prev;
}

// Function to check if linked list is palindrome
int isPalindrome(struct Node* head) {
    if (head == NULL || head->next == NULL)
        return 1;

    struct Node *slow = head, *fast = head;

    // Step 1: Find middle
    while (fast->next != NULL && fast->next->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
    }

    // Step 2: Reverse second half
    struct Node* secondHalf = reverse(slow->next);

    // Step 3: Compare halves
    struct Node* firstHalf = head;
    struct Node* temp = secondHalf;
    int isPalin = 1;

    while (temp != NULL) {
        if (firstHalf->data != temp->data) {
            isPalin = 0;
            break;
        }
        firstHalf = firstHalf->next;
        temp = temp->next;
    }

    // Optional: Restore the list
    slow->next = reverse(secondHalf);

    return isPalin;
}

// Function to print the list
void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

// Main
int main() {
    struct Node* head = NULL;
    int n, val;

    printf("Enter number of elements: ");
    scanf("%d", &n);

    printf("Enter the values:\n");
    for (int i = 0; i < n; i++) {
        scanf("%d", &val);
        append(&head, val);
    }

    printf("Linked List: ");
    printList(head);

    if (isPalindrome(head))
        printf("The linked list is a palindrome.\n");
    else
        printf("The linked list is not a palindrome.\n");

    return 0;
}
